### PR TITLE
fix(privval): CV ignore duplicate privval listen when connected

### PR DESCRIPTION
**Pull request by chillyvee**
Resolve https://github.com/cometbft/cometbft/issues/3828

On cometbft startup, duplicate privval request can enter the channel between the time listen begins and a send is attempted.  

The duplicate listen request will timeout since there is not another signer connecting, filling logs with errors.  A duplicate listen request is undesired.

---

#### PR checklist

- [x] Tests written/updated
- [x] Changelog entry added in `.changelog` (we use [unclog](https://github.com/informalsystems/unclog) to manage our changelog)
- [ ] Updated relevant documentation (`docs/` or `spec/`) and code comments

### CODE REVIEW

**Comment by chillyvee** [comment]
The automated test failures are interesting.  In seda mainnet, the daemon works well with the updates.  

Is there any chance the automated tests rely on the availability of a privval listener even when the remote signer is already connected?

**Comment by chillyvee** [comment]
Also there is some concern since GetAvailableConnection by consuming connectionAvailableCh

```
func (se *signerEndpoint) GetAvailableConnection(connectionAvailableCh chan net.Conn) bool {
        select {
        case se.conn = <-connectionAvailableCh:
                return true
...
}        
```

Which is only "refilled"  by service loop.  Would this introduce a blocking condition service loop does not continuously fail and repopulate the connectionAvailableCh channel?

```
func (sl *SignerListenerEndpoint) serviceLoop() {
                        // We have a good connection, wait for someone that needs one otherwise cancellation
                        sl.Logger.Info("SignerListener: Connected")
                        select {
                        case sl.connectionAvailableCh <- conn:
                        case <-sl.Quit():
                                return
                        }
```

**Comment by chillyvee** [comment]
Deadloack appears to be fixed by limiting mutex lock to variable assignment only

File: privval/signer_listener_endpoint.go [hunk 1]
```diff
@@ -181,6 +181,14 @@ func (sl *SignerListenerEndpoint) serviceLoop() {
 	for {
 		select {
 		case <-sl.connectRequestCh:
+			// On start, listen timeouts can queue a duplicate connect request to queue
+			// while the first request connects.  Drop duplicate request.
+			if sl.IsConnected() {
+				sl.Logger.Error("SignerListener: Connected. Drop Listen Request")
```
Comment (melekes) [inline privval/signer_listener_endpoint.go:187]:
```suggestion
				sl.Logger.Debug("SignerListener: Connected. Drop Listen Request")
```

File: privval/signer_endpoint.go [hunk 2]
```diff
@@ -53,12 +53,14 @@ func (se *signerEndpoint) GetAvailableConnection(connectionAvailableCh chan net.
...
-
+	// serviceLoop will deadlock if IsConnected() is called while WaitConnection() holds the mutex
 	select {
-	case se.conn = <-connectionAvailableCh:
-	case <-time.After(maxWait):
+	case conn := <-connectionAvailableCh:
+		// Lock only to modify se.conn
+		se.connMtx.Lock()
+		defer se.connMtx.Unlock()
+		se.conn = conn
```
Comment (melekes) [inline privval/signer_endpoint.go:62]:
```suggestion
		se.conn = conn
		se.connMtx.Unlock()
```

File: privval/signer_endpoint.go [hunk 3]
```diff
@@ -53,12 +53,14 @@ func (se *signerEndpoint) GetAvailableConnection(connectionAvailableCh chan net.
...
-	defer se.connMtx.Unlock()
-
+	// serviceLoop will deadlock if IsConnected() is called while WaitConnection() holds the mutex
 	select {
-	case se.conn = <-connectionAvailableCh:
-	case <-time.After(maxWait):
+	case conn := <-connectionAvailableCh:
+		// Lock only to modify se.conn
+		se.connMtx.Lock()
+		defer se.connMtx.Unlock()
```
Comment (melekes) [inline privval/signer_endpoint.go:61]:
```suggestion
```

**Review by melekes — COMMENTED** [review]
Thanks @chillyvee ❤️

**Review by melekes — APPROVED** [review]

**Comment by chillyvee** [comment]
Thank you for your code suggestions @melekes

File: privval/signer_endpoint.go [hunk 4]
```diff
@@ -53,12 +53,14 @@ func (se *signerEndpoint) GetAvailableConnection(connectionAvailableCh chan net.
...
 func (se *signerEndpoint) WaitConnection(connectionAvailableCh chan net.Conn, maxWait time.Duration) error {
-	se.connMtx.Lock()
-	defer se.connMtx.Unlock()
-
+	// serviceLoop will deadlock if IsConnected() is called while WaitConnection() holds the mutex
 	select {
-	case se.conn = <-connectionAvailableCh:
-	case <-time.After(maxWait):
+	case conn := <-connectionAvailableCh:
+		// Lock only to modify se.conn
```
Comment (cason) [inline privval/signer_endpoint.go:59]: This does the same as `SetConnection()`. We should use it instead.

**Review by cason — COMMENTED** [review]
I had really bad time trying to understand this code. Why do we keep locks for all calls, including calls with waiting involved (like the one fixed), with block read and writes involved?

Anyway, I would really appreciate a test unit producing the reported situation and a fix in the code that specifically addresses the test unit.

**Comment by chillyvee** [comment]
> Why do we keep locks for all calls, including calls with waiting involved (like the one fixed), with block read and writes involved?

@cason Appreciate your review.

The locks prevent go from reporting race conditions on asynchronous reads and writes to the same variable in different threads.

We'll try to contribute a unit test and may need to ask for help in creating it.

**Comment by chillyvee** [comment]
Test case added and checks pass

File: privval/signer_listener_endpoint.go [hunk 5]
```diff
@@ -34,11 +34,13 @@ type SignerListenerEndpoint struct {
 	connectRequestCh      chan struct{}
 	connectionAvailableCh chan net.Conn
 
-	timeoutAccept time.Duration
-	pingTimer     *time.Ticker
-	pingInterval  time.Duration
+	timeoutAccept   time.Duration
+	acceptFailCount int
```
Comment (cason) [inline privval/signer_listener_endpoint.go:38]:
```suggestion

```

File: privval/signer_listener_endpoint.go [hunk 6]
```diff
@@ -34,11 +34,13 @@ type SignerListenerEndpoint struct {
...
+	pingTimer       *time.Ticker
+	pingInterval    time.Duration
 
 	instanceMtx cmtsync.Mutex // Ensures instance public methods access, i.e. SendRequest
+	ivarMtx     cmtsync.Mutex // Ensures instance variable access, i.e. acceptFailCount
```
Comment (cason) [inline privval/signer_listener_endpoint.go:43]:
```suggestion
        acceptFailCount int
	ivarMtx                cmtsync.Mutex // Ensures instance variable access, i.e. acceptFailCount
```

Comment (cason) [reply privval/signer_listener_endpoint.go:43]: It is probably enough to have an atomic integer here, but works the same way with that.

File: privval/signer_listener_endpoint.go [hunk 7]
```diff
@@ -181,9 +195,17 @@ func (sl *SignerListenerEndpoint) serviceLoop() {
...
+			if sl.IsConnected() {
+				sl.Logger.Debug("SignerListener: Connected. Drop Listen Request")
+				continue
+			}
+
+			// Listen for remote signer
 			conn, err := sl.acceptNewConnection()
 			if err != nil {
-				sl.Logger.Error("SignerListener: Error accepting connection", "err", err)
+				sl.Logger.Error("SignerListener: Error accepting connection", "err", err, "acceptFailCount", sl.getAcceptFailCount())
```
Comment (cason) [inline privval/signer_listener_endpoint.go:208]:
```suggestion
				sl.Logger.Error("SignerListener: Error accepting connection", "err", err, "failures", sl.getAcceptFailCount())
```

**Review by cason — APPROVED** [review]
Thank you for adding the test unit, identifying and fixing the behavior :)

File: privval/signer_listener_endpoint.go (hunk 6 above)
Comment (melekes) [reply privval/signer_listener_endpoint.go:43]: I'm in favor of atomic int as well

Comment (chillyvee) [reply privval/signer_listener_endpoint.go]: changed to atomic.Uint32

File: privval/signer_listener_endpoint.go (hunk 7 above)
Comment (chillyvee) [reply privval/signer_listener_endpoint.go]: updated

**Comment by chillyvee** [comment]
Checks pass with atomic uint32

**Comment by cason** [comment]
I think we should merge this once tests are ok (they should, the problems are due to concurrently changes on `main`). What do you think, @melekes ?

**Comment by chillyvee** [comment]
All checks passing except for this:

https://github.com/cometbft/cometbft/pull/3830/checks?check_run_id=29365796524
```
Mergify / Rule: Make sure PR are up to date before merging (update)
failed 3 hours ago in 0s
Mergify doesn't have permission to update
For security reasons, Mergify can't update this pull request. Try updating locally.
GitHub response: refusing to allow a GitHub App to create or update workflow .github/workflows/e2e-manual-debug.yml without workflows permission
```

Is there anything we can do to assist?  Not familiar with this exact check, but doesn't look like there are any conflicts.

**Pull request merged**